File: src/main/java/com/arnav/tutionSAAS/service/AttendanceService.java
package com.arnav.tutionSAAS.service;

import com.arnav.tutionSAAS.dto.AttendanceRequest;
import com.arnav.tutionSAAS.dto.AttendanceResponse;
import com.arnav.tutionSAAS.entity.*;
import com.arnav.tutionSAAS.repository.AttendanceRepo;
import com.arnav.tutionSAAS.repository.ClassSessionRepo;
import com.arnav.tutionSAAS.repository.UserRepo;
import com.arnav.tutionSAAS.util.AttendanceMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

@Service
public class AttendanceService {

    @Autowired
    private AttendanceRepo attendanceRepo;

    @Autowired
    private ClassSessionRepo sessionRepo;

    @Autowired
    private UserRepo userRepo;

    @Autowired
    private AttendanceMapper attendanceMapper;

    @Transactional
    public AttendanceResponse markAttendance(Long sessionId, AttendanceRequest request, String clerkId) {
        ClassSession session = sessionRepo.findById(sessionId)
                .orElseThrow(() -> new RuntimeException("Session not found"));

        // Verify the teacher owns this session's batch
        User teacher = userRepo.findByClerkId(clerkId)
                .orElseThrow(() -> new RuntimeException("Teacher not found"));
        if (!session.getBatch().getTeacher().getId().equals(teacher.getId())) {
            throw new RuntimeException("You can only mark attendance for your own sessions");
        }

        List<AttendanceRecord> records = new ArrayList<>();
        for (AttendanceRequest.AttendanceEntry entry : request.getEntries()) {
            User student = userRepo.findById(entry.getStudentId())
                    .orElseThrow(() -> new RuntimeException("Student not found: " + entry.getStudentId()));

            AttendanceRecord record = attendanceMapper.toAttendanceRecord(session, student, entry.isPresent());
            records.add(record);
        }

        List<AttendanceRecord> saved = attendanceRepo.saveAll(records);
        return attendanceMapper.toAttendanceResponse(session, saved);
    }

    public AttendanceResponse getAttendanceForSession(Long sessionId) {
        ClassSession session = sessionRepo.findById(sessionId)
                .orElseThrow(() -> new RuntimeException("Session not found"));

        List<AttendanceRecord> records = attendanceRepo.findBySession_Id(sessionId);
        return attendanceMapper.toAttendanceResponse(session, records);
    }

    public AttendanceSummary getStudentAttendanceSummary(Long batchId, Long studentId) {
        long total = attendanceRepo.countByStudent_IdAndSession_Batch_Id(studentId, batchId);
        long attended = attendanceRepo.countByStudent_IdAndPresentTrueAndSession_Batch_Id(studentId, batchId);

        AttendanceSummary summary = new AttendanceSummary();
        summary.setTotalSessions(total);
        summary.setAttendedSessions(attended);
        summary.setPercentage(total > 0 ? (double) attended / total * 100.0 : 0.0);
        return summary;
    }

    // Inner class for attendance summary
    @lombok.Data
    public static class AttendanceSummary {
        private long totalSessions;
        private long attendedSessions;
        private double percentage;
    }
}

File: backend-nest/src/attendance/attendance.service.ts
import {
  ForbiddenException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { Role } from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';

@Injectable()
export class AttendanceService {
  constructor(private prisma: PrismaService) {}

  async markAttendance(
    sessionId: number,
    dto: { entries: { studentId: number; present: boolean }[] },
    clerkId: string,
  ) {
    const session = await this.prisma.classSession.findUnique({
      where: { id: sessionId },
      include: { batch: { include: { teacher: true } } },
    });
    if (!session) throw new NotFoundException('Session not found');

    const teacher = await this.prisma.user.findUnique({ where: { clerkId } });
    if (!teacher) throw new NotFoundException('Teacher not found');

    if (session.batch.teacher.id !== teacher.id) {
      throw new ForbiddenException('You can only mark attendance for your own sessions');
    }

    for (const entry of dto.entries) {
      await this.prisma.attendanceRecord.upsert({
        where: { sessionId_studentId: { sessionId, studentId: entry.studentId } },
        update: { present: entry.present, markedAt: new Date() },
        create: { sessionId, studentId: entry.studentId, present: entry.present },
      });
    }

    const records = await this.prisma.attendanceRecord.findMany({
      where: { sessionId },
      include: { student: true },
    });

    return toAttendanceResponse(session, records);
  }

  async getAttendanceForSession(sessionId: number) {
    const session = await this.prisma.classSession.findUnique({
      where: { id: sessionId },
    });
    if (!session) throw new NotFoundException('Session not found');

    const records = await this.prisma.attendanceRecord.findMany({
      where: { sessionId },
      include: { student: true },
    });

    return toAttendanceResponse(session, records);
  }

  async getStudentAttendanceSummary(
    batchId: number,
    studentId: number,
    requestingUser: any,
    clerkId: string,
  ) {
    if (requestingUser.role === Role.STUDENT) {
      if (requestingUser.id !== studentId) {
        throw new ForbiddenException('You can only view your own attendance');
      }
    } else if (requestingUser.role === Role.TEACHER) {
      const batch = await this.prisma.batch.findUnique({ where: { id: batchId } });
      if (!batch) throw new NotFoundException('Batch not found');
      const teacher = await this.prisma.user.findUnique({ where: { clerkId } });
      if (!teacher || batch.teacherId !== teacher.id) {
        throw new ForbiddenException('You can only view attendance for your own batches');
      }
    }

    const total = await this.prisma.attendanceRecord.count({
      where: { studentId, session: { batchId } },
    });
    const attended = await this.prisma.attendanceRecord.count({
      where: { studentId, present: true, session: { batchId } },
    });

    return {
      totalSessions: total,
      attendedSessions: attended,
      percentage: total > 0 ? (attended / total) * 100 : 0,
    };
  }
}

function toAttendanceResponse(session: any, records: any[]) {
  return {
    sessionId: session.id,
    sessionTitle: session.title,
    entries: records.map((r) => ({
      studentId: r.studentId,
      studentName: r.student?.fullName || r.student?.email,
      present: r.present,
      markedAt: r.markedAt,
    })),
  };
}
